tonic/src/codec/compression.rs

```diff
@@ -0,0 +1,189 @@
+use super::encode::BUFFER_SIZE;
+use crate::{metadata::MetadataValue, Status};
+use bytes::{Buf, BufMut, BytesMut};
+use flate2::read::{GzDecoder, GzEncoder};
+use std::fmt;
+
+pub(crate) const ENCODING_HEADER: &str = "grpc-encoding";
+pub(crate) const ACCEPT_ENCODING_HEADER: &str = "grpc-accept-encoding";
+
+/// Struct used to configure which encodings are enabled on a server or channel.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct EnabledCompressionEncodings {
+    pub(crate) gzip: bool,
+}
+
+impl EnabledCompressionEncodings {
+    /// Check if `gzip` compression is enabled.
+    pub fn gzip(self) -> bool {
+        self.gzip
+    }
+
+    /// Enable `gzip` compression.
+    pub fn enable_gzip(&mut self) {
+        self.gzip = true;
+    }
+
+    pub(crate) fn into_accept_encoding_header_value(self) -> Option<http::HeaderValue> {
+        let Self { gzip } = self;
+        if gzip {
+            Some(http::HeaderValue::from_static("gzip,identity"))
+        } else {
+            None
+        }
+    }
+}
+
+/// The compression encodings Tonic supports.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum CompressionEncoding {
+    #[allow(missing_docs)]
+    Gzip,
+}
+
+impl CompressionEncoding {
+    /// Based on the `grpc-accept-encoding` header, pick an encoding to use.
+    pub(crate) fn from_accept_encoding_header(
+        map: &http::HeaderMap,
+        enabled_encodings: EnabledCompressionEncodings,
+    ) -> Option<Self> {
+        let header_value = map.get(ACCEPT_ENCODING_HEADER)?;
+        let header_value_str = header_value.to_str().ok()?;
+
+        let EnabledCompressionEncodings { gzip } = enabled_encodings;
+
+        split_by_comma(header_value_str).find_map(|value| match value {
+            "gzip" if gzip => Some(CompressionEncoding::Gzip),
+            _ => None,
+        })
+    }
+
+    /// Get the value of `grpc-encoding` header. Returns an error if the encoding isn't supported.
+    pub(crate) fn from_encoding_header(
+        map: &http::HeaderMap,
+        enabled_encodings: EnabledCompressionEncodings,
+    ) -> Result<Option<Self>, Status> {
+        let header_value = if let Some(value) = map.get(ENCODING_HEADER) {
+            value
+        } else {
+            return Ok(None);
+        };
+
+        let header_value_str = if let Ok(value) = header_value.to_str() {
+            value
+        } else {
+            return Ok(None);
+        };
+
+        let EnabledCompressionEncodings { gzip } = enabled_encodings;
+
+        match header_value_str {
+            "gzip" if gzip => Ok(Some(CompressionEncoding::Gzip)),
+            other => {
+                let mut status = Status::unimplemented(format!(
```

Comment:
is this the correct status other implementations return?

Reply:
Yes:

> If a client message is compressed by an algorithm that is not supported by a server, the message WILL result in an UNIMPLEMENTED error status on the server. The server will then include a grpc-accept-encoding response header which specifies the algorithms that the server accepts. If the client message is compressed using one of the algorithms from the grpc-accept-encoding header and an UNIMPLEMENTED error status is returned from the server, the cause of the error MUST NOT be related to compression. If a server sent data which is compressed by an algorithm that is not supported by the client, an INTERNAL error status will occur on the client side.

From https://github.com/grpc/grpc/blob/master/doc/compression.md